envision/data_formatter.py

```diff
@@ -0,0 +1,364 @@
+# MIT License
+#
+# Copyright (C) 2021. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from enum import IntEnum, unique
+from types import TracebackType
+from typing import (
+    Any,
+    Callable,
+    ContextManager,
+    Dict,
+    Generator,
+    Hashable,
+    Iterable,
+    Iterator,
+    List,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Set,
+    Type,
+    Union,
+)
+
+import numpy as np
+
+from envision.types import State, TrafficActorState, TrafficActorType, VehicleType
+from smarts.core.events import Events
+from smarts.core.road_map import Waypoint
+from smarts.core.utils.file import unpack
+
+
+@unique
+class Operation(IntEnum):
+    """Formatting operations that should be performed on an object or layer."""
+
+    NONE = 0
+    """No special operation. Value will be sent as is."""
+    REDUCE = 1
+    """Send value only if it has changed."""
+    FLATTEN = 4
+    """Convert value from list or dataclass to higher hierachy."""
+    OPTIONAL = 8
+    """Sending this value is togglable by option."""
+
+
+_formatter_map: Dict[Type, Callable[[Any, "EnvisionDataFormatter"], None]] = {}
+_sequence_formatter_map: Dict[Type, Callable[[Any, "EnvisionDataFormatter"], None]] = {}
+
+_primitives = {int, float, str, VehicleType, TrafficActorType}
+
+
+class ReductionContext:
+    """Mappings between an object and its reduction to an ID."""
+
+    def __init__(
+        self,
+        mapping: Optional[Dict[Hashable, int]] = None,
+        removed: Optional[List[int]] = None,
+        enabled: bool = True,
+    ) -> None:
+        self.current_id = 0
```

Comment:
Can `self.current_id` be converted to `self._current_id` as it is not accessed outside the class?